### ISSUE TITLE
Lookups

### ISSUE THREAD

**Issue by trondhindenes** [awslabs/cdk8s#516]
### Description of the feature or enhancement:
It would be good if cdk8s had built-in support to import existing values, if exist.

A use case for this would be to instantiate secrets that are then "filled" by init containers or similar: If the secret doesn't exist, create it and set `data.mykey` to an empty string. However, if it does exist, do not "reset" it, instead import the existing value if `data.mykey` and re-use that. 

In helm this is typically what the `lookup` function is used for.

### Use Case:
See description.
One could of course instantiate a regular kubernetes client "on the side" but that would potentially bloat the code. It would be better (imho) if cdk8s contained clean lookup functions.

### Other:

* [ ] :wave: I may be able to implement this feature request
* [ ] :warning: This feature might incur a breaking change

---

This is a :rocket: Feature Request

**Comment by iliapolo** [comment]
@trondhindenes We've been thinking about this and are definitely interested to hear some concrete use-cases. Can you elaborate on what exactly you had in mind? Are you looking to use a concrete secret value during synthesis? Because just *referencing* existing secrets in the manifest is supported by using `Secret.fromSecretName` and passing the `ISecret` around.

**Comment by trondhindenes** [comment]
Yup, we actually merged one in the dapr project (which is using Helm) yesterday that is a good example of this:
https://github.com/dapr/dapr/blob/master/charts/dapr/charts/dapr_sidecar_injector/templates/dapr_sidecar_injector_webhook_config.yaml

Here's what happens: 
- Helm creates the secret (without any data) if it doesnt exist
- A pod starts up, and if the secret is empty, it gets "filled"
- Upon upgrades (Helm upgrade) we read the current contents of the secret data, and "pass that back in". This lets us have logic that can "create stuff if it doesnt exist", but once there is data in the secret, that data will be retained.

The nice thing about this is that "the tool" (Helm in this case) still manages the secret object, so when removing the "app", the secret will also be deleted.

The last one is important, otherwise one could just import using the `fromSecretName` function you mentioned.

I hope that makes it a bit clearer what _my_ use case is, at least. I'm sure there are many other uses-cases for this too.

**Comment by iliapolo** [comment]
@trondhindenes Thanks for the breakdown. I'll put it on our docket and let it collect some more feedback and use-cases before we try and tackle it. 

In the meantime, I'm interested in the specifics of what you described.

> A pod starts up, and if the secret is empty, it gets "filled"

The pod waits until the secret is created right? Or are you enforcing a resource creation order somehow? 
Have you considered a model where the owner of the secret is actually the pod itself, i.e responsible for creating it, and hence for deleting it. 

Not saying its a better option, there are definitely some kinks around that as well. But interested in your thoughts. This would only work if the same secret isn't shared across different pods, so when the pod deletes the secret, it doesn't screw things up for other pods that may still be running.

**Comment by trondhindenes** [comment]
> The pod waits until the secret is created right

So, the secret exists, but it doesnt contain data. 

A valid example is generating a self-signed cert (for example used by a mutating admission webhook or similar). Pod starts, sees that the secret doesn't contain any data, so it generates a self-signed cert, and stores it in the secret. Subsequent pod restarts/upgades will simply re-use the existing cert already there.

This might be a bit of a edge case but I'd argue that it's a valid one. Another one is to be able to read the current replica count of a deployment, so that an upgrade doesn't interfere with the instance count.

These are just from the top of my head - there are many reasons for needing to "call in" to grab existing data.
